misc/pragmas.md

```diff
@@ -175,51 +172,137 @@ optional mark, a QName, the pragma's data, and a right square bracket.
-      XML representation of the symbol; the name and data of the
-      processing instruction in the XML form are the name and pragma
-      data of the pragma in the ixml form.
+These XML constructs may occur in locations corresponding to those in
+which pragmas may appear in the ixml grammar:
 
-    * serialized as extension elements if unmarked or marked `^`.
+* as a child or attribute of the `ixml` element before, between, or
+  after `rule` elements.
 
-      The extension elements appear as the first children of the XML
-      representation of the symbol. In the canonical XML
-      representation, the pragma data is serialized as a 'pragma-data'
-      element within the extension element, which will contain the
-      pragma-data as a character sequence. Depending on the semantics
-      of the pragma itself, the extension element may also contain
-      additional elements. (See the examples in the final section of
-      this paper.)
+* as a child or attribute of a `rule` element.  An ixml pragma marked 
+  `^` or `?` occurring on the left-hand side of a rule corresponds to an
+  extension element or processing instruction occurring before the
+  first `alt` child of the `rule` element; if it occurs immediately
+  before the full stop of the rule, it corresponds to an element or
+  processing instruction occurring after the last `alt` of the `rule`.
+
+* as a child or attribute of a `nonterminal`, `literal`, `inclusion`,
+  or `exclusion` element.
+
+To ensure that the ixml grammar has an XML representation, any two
+pragmas marked `@` and attached to the same construct must have
+different expanded names.
+
+When a grammar in XML form is serialized into ixml form, extension
+attributes appearing on the `rule` element may be serialized either on
+the left-hand side of the rule or before the full stop.  Pragmas
+attached to a symbol in the right-hand side of a rule may be
+serialized either before or after the mark; the two positions are
+equivalent.
+
+For attributes, the attribute name is the QName of the ixml pragma 
+and the attribute value is the pragma data. 
+
+For processing instructions, the PI name is the QName of the 
+ixml pragma and the PI value is the pragma data. 
+
+For extension elements, the element name is the QName of the ixml
+pragma and the pragma data appears as character data within a child
+named `pragma-data`.  Other child elements may follow the
+`pragma-data` element; their content is required to be reconstructible
+from the pragma data and the fallback expression, but they may express
+the information in a more convenient form.  (For example, the pragma
+data may be a structured expression which a conforming application
+will parse; the parsed form of the pragma data may be enclosed in the
+pragma.)  In this way we ensure that the ixml and XML forms of a
+grammar contain the same information, although the XML form of the
+grammar may be easier to process by machine.
```

Comment:
The draft does not enforce any data structure; it only provides that 
an extension element can contain two different representations of
the relevant information.  The first of these two representations is 
the content of the `pragma-data` element:  the same sequence of 
characters which appears (or *would* appear) between the blank
following the QName and the closing bracket in the ixml form of the pragma.
The second is implementation-defined or implementation-dependent
and is unconstrained by the pragmas proposal.  The only thing the proposal
does is tell implementors that if they have stuff they need to put somewhere
in the XML form, the extension element is the place to put it.

The proposal does need, I think, to say explicitly that the pragma data in the
ixml form of a pragma reappears as the character sequence in the `pragma-data`
element in the XML form of the pragma, and that anything else in the 
extension element on the XML side does not get serialized on the ixml side.

Saying that the XML form of a pragma may not contain anything but characters
is a way of ensuring that implementors just ignore the rules, or extend the XML
syntax since pragmas are not workable.  I will be willing to discuss a rule that
says extension elements may only contain PCDATA if someone shows me an XML version
of the grammars in the grammar-rewriting and token-annotation that I can take seriously
as something I could use.

Reply:
I think the `pragma-data` element itself is an enforced structure; I'm not against implementation-defined XML structure for a given pragma, I just don't understand what `pragma-data` brings to the party as a default representation compared to, say, a text node.

To be clear, I am not proposing that we should be mandating the structure in the vXML representation within the pragma extension element at all: I am arguing for the opposite.  The alternatives I list above were alternatives to the main paragraph, "it is at least necessary to state that the ixml and XML forms must contain the same information in a[n implementation] defined, round-trippable way".

I agree that the proposal doesn't need to explicitly state how the pragma data relates to the content of the pragma element in the vXML representation: I think that dropping `pragma-data` is coherent to that approach.

Reply:
I'm sorry, but I don't understand what you are suggesting, or why you do not see what the `pragma-data` element contributes.

If we drop the `pragma-data` element and allow arbitrary structure within the XML representation of a pragma, how does a processor that does not understand the pragma in question produce an XML form of an ixml grammar, or an ixml form of an XML grammar?

If we drop the `pragma-data` element and specify that the XML pragma contains the same sequence of characters as the corresponding ixml pragma, how do use cases like the alternative-formulation use case work?

The function of the `pragma-data` element in the proposal as drafted is to provide both inter-convertibility of XML and ixml forms of the pragma and the ability of the XML form of the pragma to contain additional redundant
structure.  If we lose the `pragma-data` element, it seems to me we lose one or the other of those properties, or both.

To take a concrete example.  In the XML form of a grammar, we have this extension element.  I have removed the `pragma-data` element, since you appear to be suggesting it's unnecessary.  
````
        <r:ref>
          <rule name="factor" mark="-">
            <alt><nonterminal name="number"/></alt>
            <alt><nonterminal name="var"/></alt>
            <alt>
              <literal sstring="(" tmark="-"/>
              <nonterminal name="EXPR" mark="-"/>
              <literal sstring="-" tmark="-"/>
            </alt>
          </rule>
        </r:ref>
````

Now please show me how to serialize this meaningfully as a pragma in ixml form.  I think it's going to start with `[r:ref ` and end with `]`.  And what comes in between?  Something? Nothing?  How do anyone know?

Or are you assuming that grammars containing a given kind of pragma can only be converted between ixml and XML forms by processors that understand that pragma?  That seems to violate the principle of pragmas 
not interfering with the operation of processors that do not understand the pragmas.

I apologize in advance.  As far as I can tell everything I have just said is just a repetition of things said earlier.  If they weren't clear then, they won't be clear now.  But I am completely at a loss as to what issue you have with the `pragma-data` element or how you think the proposal would function in practice without it.

To avoid an infinite regress of shrugs and asking "what is it you don't understand about what it is I don't understand?" I suppose I should try to say what it is I don't understand about your comment.

* You refer to `pragma-data` as "a default representation [of a pragma, in XML form] compared to, say, a text node."  

  I'm not sure where the notion of *default representation* comes from.  I don't see a default
representation in the proposal:  I see a provision that in XML form, pragmas are extension elements containing a `pragma-data` element and optionally additional elements.  If there is no `pragma-data` element, the extension element is not a pragma, as I understand the proposal.  What's "default" about that?

* You say "To be clear, I am not proposing that we should be mandating the structure in the vXML representation within the pragma extension element at all".   
 
  That does seem clear to me; the only thing that puzzles me is why you would expect anyone to misunderstand, which suggests to me I may have missed something important.  This point is probably not worth pursuing.

* You say "the proposal doesn't need to explicitly state how the pragma data relates to the content of the pragma  element in the vXML representation".  

  In that case, how does a processor that does not understand a pragma convert the grammar between ixml and XML forms?